### ISSUE TITLE
dotnet format is broken on sdk 7.0.200 if `EnforceCodeStyleInBuild` is set to `true` – Could not load file or assembly 'Microsoft.CodeAnalysis'

### ISSUE THREAD

**Issue by askazakov** [dotnet/format#1800]
step to reproduce

- create new project
```
dotnet new console
```
 - add `<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>` to .csproj

- add global.json
```
dotnet new globaljson
```

- change sdk version in global.json to "7.0.103"
- run
```
dotnet format
```

everything is ok here.

- change sdk version in global.json to "7.0.200"
- run
```
dotnet format
```

and now there are a lot of 
```
Unhandled exception: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.5.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.
```
in output

**Comment by askazakov** [comment]
```
dotnet --info
.NET SDK:
 Version:   7.0.200
 Commit:    534117727b

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  13.2
 OS Platform: Darwin
 RID:         osx.13-x64
 Base Path:   /usr/local/share/dotnet/sdk/7.0.200/

Host:
  Version:      7.0.3
  Architecture: x64
  Commit:       0a2bda10e8
```

**Comment by tidusjar** [comment]
Same is happening for me and all of my colleagues
```
.NET SDK:
 Version:   7.0.200
 Commit:    534117727b

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.200\

Host:
  Version:      7.0.3
  Architecture: x64
  Commit:       0a2bda10e8
```

**Comment by MattMyers3** [comment]
Same issue happening for my team. It also seems like there are multiple issues related to this issue with dotnet format. https://github.com/dotnet/format/issues/1785

**Comment by si618** [comment]
Thanks for the heads-up, [confirmed](https://github.com/si618/lolcat/actions/runs/4222004277) and if nothing else a good excuse to move to [Directory.Build.Props](https://github.com/si618/lolcat/commit/68489777b7eeb2f6a200e990c5c36adc9d424f7a)

**Comment by zh6335901** [comment]
I had the same problem

**Comment by m3nax** [comment]
Same here

**Comment by panicoenlaxbox** [comment]
Idem, with `dotnet --version` 7.0.200.
@si618 Have I understood that there is workaround with `Directory.Build.props`? I've tried it and it doesn't work either.
Thank you.

**Comment by si618** [comment]
> @si618 Have I understood that there is workaround with `Directory.Build.props`? I've tried it and it doesn't work either. Thank you.

Commenting out `EnforceCodeStyleInBuild` as per linked commit worked for me.

Not really a workaround per se, just skipping until fixed, but having it in `Directory.Build.props` means only changing one file.

**Comment by panicoenlaxbox** [comment]
Ok, thank you, I hope it will be fixed soon.

**Comment by RikkiGibson** [comment]
I suspect that this is caused by dotnet-format's reference to the compiler getting out of date. #1782

**Comment by glen-84** [comment]
@RikkiGibson Do you know if there are plans to assign a new maintainer for this project?

**Comment by MrcnPnkrsk** [comment]
Is there any Road Map for fixing this issue?

.NET 7.0.4 is installed automatically with Visual Studio 2022 and for the past 4 months bundled dotnet-format is unusable.
#1782 Seems to be forgotten.

**Comment by GGG-KILLER** [comment]
I've been getting this using .NET 8 in a dev container where .NET was installed using `dotnet-install.sh`.

The project it runs into an exception while trying to format is an Incremental Source Generator.

Solution: https://github.com/LorettaDevs/Loretta (devcontainer files are contained within)
Project: `Loretta.Generators.SyntaxFactsGenerator`
Format command: `dotnet format` (v8.0.405501+2cb3e68c6b9a966114572fd63f2a20d2cb54a288)
Output: https://gist.github.com/GGG-KILLER/81a9773f7c02a12a841edd331751d2ee


<details>
<summary><code>dotnet --info</code></summary>

```console
$ dotnet --info
.NET SDK:
 Version:   8.0.100-preview.2.23157.25
 Commit:    54801b2435

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  22.04
 OS Platform: Linux
 RID:         ubuntu.22.04-x64
 Base Path:   /usr/local/dotnet/current/sdk/8.0.100-preview.2.23157.25/

Host:
  Version:      8.0.0-preview.2.23128.3
  Architecture: x64
  Commit:       30b879924a

.NET SDKs installed:
  6.0.407 [/usr/local/dotnet/current/sdk]
  7.0.202 [/usr/local/dotnet/current/sdk]
  8.0.100-preview.2.23157.25 [/usr/local/dotnet/current/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.15 [/usr/local/dotnet/current/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.4 [/usr/local/dotnet/current/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-preview.2.23153.2 [/usr/local/dotnet/current/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.15 [/usr/local/dotnet/current/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.4 [/usr/local/dotnet/current/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-preview.2.23128.3 [/usr/local/dotnet/current/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```
</details>

**Comment by avidenic** [comment]
This seems to be an issue on 7.0.202 as well.

**Comment by lucasteles** [comment]
@avidenic  .NET8 also

**Comment by m3nax** [comment]
#1825 Merged

**Comment by niemyjski** [comment]
@arkalyanms when will this be fixed? This has been broken for months.

**Comment by arkalyanms** [comment]
First off, apologies for the delays in getting this fixed! We have had bit of personnel churn in this space, but we have a fix checked in on the format side now and it is queued up to catch the next SDK servicing train that is currently planned for 7.0.3xx .NET SDK. I will circle back once I have a confirmation that the fix has been released. In the interim, dropping down to 7.0.2xx to use the dotnet format would be the temporary workaround.

**Comment by arkalyanms** [comment]
> @arkalyanms I'm guessing this is not fixed in .NET 8 Preview 4? Still getting errors:

It caught the Preview 5 SDK train. For more details, the SDK insertion that took the fix into the preview under development is https://github.com/dotnet/sdk/pull/32735.

**Comment by niemyjski** [comment]
Also @arkalyanms, how come dotnet format doesn't pick the sdk listed in the global.json file? If I have .NET 6 selected it uses the latest sdk version.

**Comment by arkalyanms** [comment]
> Thanks for that! Is there any way to get automated tests for this? I think this is the second or third time dotnet format has broken completely since 6.0 for roughly the same reason.

Absolutely yes for the integration test gap that needs to be filled. @sharwell to comment on the global json support.

**Comment by sharwell** [comment]
`dotnet format` runs from a self-contained folder on disk. Whatever code `dotnet` uses to determine the binary to run for the `format` command is what determines the SDK that provides the format implementation. This seems like a question for the `dotnet` CLI about why it isn't accounting for global.json.

**Comment by JoeRobich** [comment]
MSBuildLocator should be returning the pinned SDK first, if it was installed.

https://github.com/dotnet/format/blob/c99dcfe7784e15b7f73c76625dfd7ccc9af3ed92/src/Commands/FormatCommandCommon.cs#L350-L352

**Comment by idilov** [comment]
Both this problem and [dotnet/sdk#32598](https://github.com/dotnet/sdk/issues/32598) seem to have been fixed in sdk 7.0.304.
I re-enabled EnforceCodeStyleInBuild, removed all workarounds and downgrades which I had been using for 4 months and can once again enjoy `dotnet format` as is. Hope it's not just me. Thank you!

**Comment by brminnick** [comment]
> Both this problem and https://github.com/dotnet/sdk/issues/32598 seem to have been fixed in sdk 7.0.304.

Awesome!! Yup, I can confirm that we can now use `dotnet format` again with no errors in our [.NET Community Toolkit libraries](https://github.com/CommunityToolkit)

**Comment by JoeRobich** [comment]
@brminnick Thanks for the confirmation! Closing.
